### PR TITLE
Refactor TF and TFLite implementations into their own classes/files

### PR DESCRIPTION
**Pull request by reuben**
This refactors the ModelState code which currently into a base `ModelState` class and two implementations, `TFModelState`, `TFLiteModelState`.

This greatly improves the readability of `deepspeech.cc`. I've also taken the opportunity to try and make the code style a bit more consistent. I've made all member variable names end with `_` and fixed inconsistent indentation in a few places.

This also paves the way for fixing the multi-stream bugs we currently have in the client. Not sure if we should take this for 0.5.0 though, as it's a big change.

### CODE REVIEW

**Comment by reuben** [comment]
Hm, looks like something is wrong with Taskcluster, it didn't pick up the PR.

**Comment by lissyx** [comment]
> Hm, looks like something is wrong with Taskcluster, it didn't pick up the PR.

Retriggered them, it looks okay

File: native_client/deepspeech.cc [hunk 1]
```diff
@@ -667,182 +276,23 @@ DS_CreateModel(const char* aModelPath,
...
-  model->fbmodel = tflite::FlatBufferModel::BuildFromFile(aModelPath);
-  if (!model->fbmodel) {
-    std::cerr << "Error at reading model file " << aModelPath << std::endl;
-    return DS_ERR_FAIL_INIT_MMAP;
-  }
-
+  std::unique_ptr<ModelState> model(new TFModelState());
+#else
+  std::unique_ptr<ModelState> model(new TFLiteModelState());
+#endif // USE_TFLITE
```
Comment (lissyx) [inline native_client/deepspeech.cc]: nit: would be even simpler to read written as:
```
std::unique_ptr<ModelState> model(
#ifndef USE_TFLITE
    new TFModelState()
#else
    new TFLiteModelState()
#endif
);
```

File: native_client/tflitemodelstate.cc [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  const int final_dim_size = dims_logits->data[1] - 1;
+  if (final_dim_size != alphabet_->GetSize()) {
+    std::cerr << "Error: Alphabet size does not match loaded model: alphabet "
+              << "has size " << alphabet_->GetSize()
+              << ", but model has " << final_dim_size
+              << " classes in its output. Make sure you're passing an alphabet "
+              << "file with the same size as the one used for training."
+              << std::endl;
+    return DS_ERR_INVALID_ALPHABET;
+  }
```
Comment (lissyx) [inline native_client/tflitemodelstate.cc]: nit: maybe this kind of check should be factorized on `ModelState`, like with a `ensure_alphabet(size_t final_dim_size)` that would check against `GetSize()` and return the same error, since the logic is the same for TensorFlow and TFLite runtime, only where we read the value differs ?

File: native_client/tflitemodelstate.cc [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    }
+  }
+
+  assert(previous_state_size_ > 0);
+
+  // Feeding previous_state_c, previous_state_h
+  memcpy(interpreter_->typed_tensor<float>(previous_state_c_idx_), previous_state_c_.get(), sizeof(float) * previous_state_size_);
+  memcpy(interpreter_->typed_tensor<float>(previous_state_h_idx_), previous_state_h_.get(), sizeof(float) * previous_state_size_);
+
+  interpreter_->SetExecutionPlan(acoustic_exec_plan_);
```
Comment (lissyx) [inline native_client/tflitemodelstate.cc]: nit: I think we should check the return status here.

File: native_client/tflitemodelstate.cc [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+TFLiteModelState::TFLiteModelState()
+  : ModelState()
+  , interpreter_(nullptr)
+  , fbmodel_(nullptr)
+  , previous_state_size_(0)
+  , previous_state_c_(nullptr)
+  , previous_state_h_(nullptr)
+{
+}
```
Comment (lissyx) [inline native_client/tflitemodelstate.cc]: No `~TFLiteModelState` ? We're allocating stuff, aren't we?

Comment (lissyx) [reply native_client/tflitemodelstate.cc]: Ok, that makes sense, we `delete` decoder state in `~StreamingState`

File: native_client/tflitemodelstate.cc (hunk 2 above)
Comment (reuben) [reply native_client/tflitemodelstate.cc]: Yeah, I want to do that, ideally we'd uplift all the metadata/model geometry checks to `ModelState`, but that requires finding a good solution for embedding metadata in the TFLite model first.

**Comment by reuben** [comment]
Pushed a commit fixing multiple concurrent streams per model instance.

**Comment by lissyx** [comment]
@reuben I guess we can merge that after you rebase, now that 0.5.0 is shipped.

**Comment by reuben** [comment]
Re-requesting review for the last commit fixing the concurrent streams bug. I'm also going to add a test that exercises this.

File: DeepSpeech.py [hunk 5]
```diff
@@ -618,51 +614,31 @@ def create_inference_graph(batch_size=1, n_steps=16, tflite=False):
...
+    new_state_h = tf.identity(new_state_h, name='new_state_h')
+
+    inputs = {
+        'input': input_tensor,
+        'previous_state_c': previous_state_c,
+        'previous_state_h': previous_state_h,
+        'input_samples': input_samples,
+    }
+
+    if FLAGS.use_seq_length:
```
Comment (lissyx) [inline DeepSpeech.py]: I've been thinking about that and I'm wondering, given the reports (myself included) whether we should not also `or not FLAGS.export_tflite` so that passing `--export_tflite` is the only requirement.

File: native_client/BUILD [hunk 6]
```diff
@@ -103,34 +114,26 @@ tf_cc_shared_object(
             ### => Trying to be more fine-grained
             ### Use bin/ops_in_graph.py to list all the ops used by a frozen graph.
             ### CPU only build, libdeepspeech.so file size reduced by ~50%
-            "//tensorflow/core/kernels:dense_update_ops",   # Assign
-            "//tensorflow/core/kernels:constant_op",        # Const
-            "//tensorflow/core/kernels:immutable_constant_op", # ImmutableConst
+            "//tensorflow/core/kernels:dense_update_ops",   # Assign (remove once prod model no longer depends on it)
```
Comment (lissyx) [inline native_client/BUILD]: Hm, you removed all the `assign` op, are you referring to the prod model we release / host for running tests ?
If so, I do worry that it's a bit burried and we may forgot

File: native_client/BUILD [hunk 7]
```diff
@@ -103,34 +114,26 @@ tf_cc_shared_object(
...
             "//tensorflow/core/kernels:tile_ops",           # Tile
-            "//tensorflow/core/kernels:gather_op",          # Gather
             "//tensorflow/core/kernels:mfcc_op",            # Mfcc
             "//tensorflow/core/kernels:spectrogram_op",     # AudioSpectrogram
             "//tensorflow/core/kernels:strided_slice_op",   # StridedSlice
             "//tensorflow/core/kernels:slice_op",           # Slice, needed by StridedSlice
             "//tensorflow/contrib/rnn:lstm_ops_kernels",    # BlockLSTM
-            "//tensorflow/core/kernels:random_ops",         # RandomGammaGrad
             "//tensorflow/core/kernels:pack_op",            # Pack
             "//tensorflow/core/kernels:gather_nd_op",       # GatherNd
```
Comment (lissyx) [inline native_client/BUILD]: Have you checked on CUDA runtime ? I remember some of those dependencies were actually only coming from CUDA version.

File: native_client/tflitemodelstate.cc [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+  TfLiteIntArray* dims_h = interpreter_->tensor(previous_state_h_idx_)->dims;
+  assert(dims_c->data[1] == dims_h->data[1]);
+  assert(state_size_ > 0);
+  state_size_ = dims_c->data[1];
+
+  return DS_ERR_OK;
+}
+
+void
+TFLiteModelState::copy_vector_to_tensor(const vector<float>& vec,
```
Comment (lissyx) [inline native_client/tflitemodelstate.cc]: Is it better than just `memcpy` ?

File: native_client/tflitemodelstate.cc [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+  for (i = 0; i < vec.size(); ++i) {
+    tensor[i] = vec[i];
+  }
+  for (; i < num_elements; ++i) {
+    tensor[i] = 0.f;
+  }
+}
+
+void
+TFLiteModelState::copy_tensor_to_vector(int tensor_idx,
```
Comment (lissyx) [inline native_client/tflitemodelstate.cc]: Same question

File: native_client/tflitemodelstate.cc [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+                        vector<float>& state_c_output,
+                        vector<float>& state_h_output)
+{
+  const size_t num_classes = alphabet_->GetSize() + 1; // +1 for blank
+
+  // Feeding input_node
+  copy_vector_to_tensor(mfcc, input_node_idx_, n_frames*mfcc_feats_per_timestep_);
+
+  // Feeding previous_state_c, previous_state_h
+  assert(previous_state_c.size() == state_size_);
```
Comment (lissyx) [inline native_client/tflitemodelstate.cc]: maybe assert within the copy functions ?

File: native_client/tflitemodelstate.cc (hunk 8 above)
Comment (lissyx) [reply native_client/tflitemodelstate.cc]: Ok, from IRC, this is for zero padding mostly. Might be worth putting a comment.

**Comment by lissyx** [comment]
@reuben I landed a fix for the ElectronJS build issues, you can rebase on top of it.

File: native_client/BUILD (hunk 6 above)
Comment (reuben) [reply native_client/BUILD]: Yep. If we remove this then prod model tests fail when loading the graph.

File: native_client/tflitemodelstate.cc (hunk 9 above)
Comment (reuben) [reply native_client/tflitemodelstate.cc]: Can't memcpy into an std::vector. This could be implemented as an std::copy call though, if you prefer.

File: native_client/tflitemodelstate.cc (hunk 10 above)
Comment (reuben) [reply native_client/tflitemodelstate.cc]: The copy functions are agnostic to what tensor you're copying stuff into, and they're meant to work with different size vector and `num_elements`, so I can't put these asserts there.

File: native_client/tflitemodelstate.cc (hunk 9 above)
Comment (lissyx) [reply native_client/tflitemodelstate.cc]: I guess any implem is fine, I don't think it's going to be a big bottleneck on the perfs

File: native_client/BUILD (hunk 6 above)
Comment (lissyx) [reply native_client/BUILD]: Maybe we should file an issue to not forget and refer to the issue here?

Comment (reuben) [reply native_client/BUILD]: Filed #2179

**Comment by reuben** [comment]
@lissyx I've addressed your comments and also added a test in f12ea5e. The test uses the prod model so it's not actually running, since the graph version was bumped here, but I verified manually. I'll also re-export the 0.5.0 checkpoint and update our prod-model in a separate PR.

**Comment by reuben** [comment]
I just saw your messages about 0.5.1 with the DS_IntermediateDecode fix. I'll hold on merging this PR until we do that.
